packages/cli/src/lib/bundler/config.ts

```diff
@@ -232,12 +233,47 @@ export async function createConfig(
     require.resolve('react-refresh'),
   ];
 
+  const mode = isDev ? 'development' : 'production';
+  const optimization = optimizationConfig(options);
+
+  if (process.env.FORCE_REACT_DEVELOPMENT === 'true') {
+    console.log(
+      chalk.yellow(
+        `⚠️  WARNING: Forcing react and react-dom into development mode. This build should not be used in production.`,
+      ),
+    );
+
+    const reactPackageDirs = [
+      `${dirname(require.resolve('react/package.json'))}/`,
+      `${dirname(require.resolve('react-dom/package.json'))}/`,
+    ];
+
+    // Don't define process.env.NODE_ENV with value matching config.mode.
+    optimization.nodeEnv = false;
+
+    // Instead, provide a custom definition which always uses "development" if
+    // the module is part of `react` or `react-dom`, and `config.mode` otherwise.
+    plugins.push(
+      new webpack.DefinePlugin({
+        'process.env.NODE_ENV': webpack.DefinePlugin.runtimeValue(
+          ({ module }) => {
+            if (reactPackageDirs.some(val => module.resource.startsWith(val))) {
+              return '"development"';
+            }
+
+            return `"${mode}"`;
+          },
+        ),
+      }),
+    );
+  }
+
   const withCache = yn(process.env[BUILD_CACHE_ENV_VAR], { default: false });
 
   return {
-    mode: isDev ? 'development' : 'production',
+    mode,
     profile: false,
-    optimization: optimization(options),
+    optimization,
```

Comment:
Small refactor down here to allow keeping the config related to forcing react into development mode in one place.